File: Backend/routes/invitations.py
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity, get_jwt
from models import db
from models.invitation import Invitation
from models.brand import Brand
from models.influencer import Influencer
from models.campaign import Campaign

invitations_bp = Blueprint('invitations', __name__)


@invitations_bp.route('', methods=['GET'])
@jwt_required()
def list_invitations():
    user_id = get_jwt_identity()
    claims = get_jwt()

    if claims.get('role') == 'influencer':
        influencer = Influencer.query.filter_by(user_id=user_id).first()
        if not influencer:
            return jsonify(success=True, data=[]), 200
        invitations = Invitation.query.filter_by(influencer_id=influencer.id)\
            .order_by(Invitation.created_at.desc()).all()
    else:
        brand = Brand.query.filter_by(user_id=user_id).first()
        if not brand:
            return jsonify(success=True, data=[]), 200
        invitations = Invitation.query.filter_by(brand_id=brand.id)\
            .order_by(Invitation.created_at.desc()).all()

    return jsonify(success=True, data=[i.to_dict() for i in invitations]), 200


@invitations_bp.route('', methods=['POST'])
@jwt_required()
def create_invitation():
    claims = get_jwt()
    if claims.get('role') != 'brand':
        return jsonify(success=False, error='Brand role required'), 403

    data = request.get_json()
    user_id = get_jwt_identity()
    brand = Brand.query.filter_by(user_id=user_id).first()
    if not brand:
        return jsonify(success=False, error='Brand not found'), 404

    campaign = Campaign.query.get(data.get('campaignId'))
    if not campaign:
        return jsonify(success=False, error='Campaign not found'), 404

    influencer = Influencer.query.get(data.get('influencerId'))
    if not influencer:
        return jsonify(success=False, error='Influencer not found'), 404

    invitation = Invitation(
        campaign_id=campaign.id,
        campaign_title=campaign.title,
        brand_id=brand.id,
        brand_name=brand.company_name,
        influencer_id=influencer.id,
        influencer_name=influencer.display_name,
        message=data.get('message', ''),
        status='pending',
    )
    db.session.add(invitation)
    db.session.commit()
    return jsonify(success=True, data=invitation.to_dict()), 201


@invitations_bp.route('/<invitation_id>/status', methods=['PUT'])
@jwt_required()
def update_status(invitation_id):
    data = request.get_json()
    new_status = data.get('status')
    if new_status not in ('accepted', 'declined'):
        return jsonify(success=False, error='Status must be accepted or declined'), 400

    invitation = Invitation.query.get(invitation_id)
    if not invitation:
        return jsonify(success=False, error='Invitation not found'), 404

    invitation.status = new_status
    db.session.commit()
    return jsonify(success=True, message=f'Invitation {new_status}'), 200

File: server/src/routes/invitations.ts
import { Router, Request, Response, NextFunction } from 'express';
import { eq } from 'drizzle-orm';
import { v4 as uuid } from 'uuid';
import db from '../db/index.js';
import { invitations, brands, influencers, campaigns } from '../db/schema.js';
import { authenticate, authorize } from '../middleware/auth.js';
import { AppError } from '../middleware/error.js';
import { sendEmail, campaignInviteEmail } from '../lib/email.js';

const router = Router();

router.get('/', authenticate, async (req: Request, res: Response, next: NextFunction) => {
  try {
    let results;
    if (req.user!.role === 'influencer') {
      const inf = await db.select().from(influencers).where(eq(influencers.userId, req.user!.sub)).limit(1);
      results = inf.length ? await db.select().from(invitations).where(eq(invitations.influencerId, inf[0].id)) : [];
    } else {
      const brand = await db.select().from(brands).where(eq(brands.userId, req.user!.sub)).limit(1);
      results = brand.length ? await db.select().from(invitations).where(eq(invitations.brandId, brand[0].id)) : [];
    }
    res.json({ success: true, data: results });
  } catch (err) { next(err); }
});

router.post('/', authenticate, authorize('brand'), async (req: Request, res: Response, next: NextFunction) => {
  try {
    const { campaignId, influencerId, message } = req.body;
    const brand = await db.select().from(brands).where(eq(brands.userId, req.user!.sub)).limit(1);
    if (!brand.length) throw new AppError('Brand not found', 404);
    const campaign = await db.select().from(campaigns).where(eq(campaigns.id, campaignId)).limit(1);
    if (!campaign.length) throw new AppError('Campaign not found', 404);
    const inf = await db.select().from(influencers).where(eq(influencers.id, influencerId)).limit(1);
    if (!inf.length) throw new AppError('Influencer not found', 404);

    const id = uuid();
    await db.insert(invitations).values({
      id, campaignId, campaignTitle: campaign[0].title,
      brandId: brand[0].id, brandName: brand[0].companyName,
      influencerId, influencerName: inf[0].displayName,
      message: message || '', status: 'pending',
    });

    // Send email notification
    const emailContent = campaignInviteEmail(inf[0].displayName, brand[0].companyName, campaign[0].title);
    sendEmail({ to: inf[0].email, ...emailContent }).catch(() => {});

    res.status(201).json({ success: true, message: 'Invitation sent' });
  } catch (err) { next(err); }
});

router.patch('/:id/status', authenticate, async (req: Request, res: Response, next: NextFunction) => {
  try {
    const { status } = req.body;
    if (!['accepted', 'declined'].includes(status)) throw new AppError('Invalid status', 400);
    await db.update(invitations).set({ status }).where(eq(invitations.id, req.params.id));
    res.json({ success: true, message: `Invitation ${status}` });
  } catch (err) { next(err); }
});

export default router;
